File: packages/sdk-py/src/flexvaults/auth/siwe.py
from __future__ import annotations

import asyncio
import secrets
import string
import weakref
from dataclasses import dataclass
from datetime import datetime, timezone
from time import monotonic_ns
from typing import Protocol

from eth_account.messages import encode_defunct
from eth_account.signers.local import LocalAccount

from ..types import Address, HexString, normalize_address, normalize_hex
from ..types.responses import SiweDomainResponse, SiweLoginResponse


class SiweSigner(Protocol):
    async def __call__(self, message: str) -> HexString: ...


class SiweAuthClient(Protocol):
    @property
    def base_url(self) -> str: ...

    def set_siwe_token(self, token: str) -> None: ...

    async def get_siwe_domain(self) -> SiweDomainResponse: ...

    async def siwe_login(self, *, siwe_message: str, signature: str) -> SiweLoginResponse: ...


def _random_nonce(length: int = 8) -> str:
    charset = string.ascii_letters + string.digits
    return "".join(secrets.choice(charset) for _ in range(length))


def build_siwe_message(
    *,
    domain: str,
    address: Address | str,
    chain_id: int,
    statement: str = "Sign in to Flexvaults",
    uri: str | None = None,
    nonce: str | None = None,
    issued_at: str | datetime | None = None,
) -> str:
    """
    Build a minimal SIWE message (EIP-4361) compatible with AccountingSiweAuth.

    Matches the TypeScript SDK format:
    - No expiration time (contract defaults to 24h)
    - No resources (informational only on-chain currently)
    """
    normalized_domain = domain.strip()
    normalized_address = normalize_address(str(address))
    normalized_chain_id = int(chain_id)
    normalized_uri = uri or f"https://{normalized_domain}"
    normalized_nonce = nonce or _random_nonce()
    if issued_at is None:
        normalized_issued_at = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    elif isinstance(issued_at, datetime):
        normalized_issued_at = issued_at.astimezone(timezone.utc).isoformat().replace("+00:00", "Z")
    else:
        normalized_issued_at = issued_at

    return "\n".join(
        [
            f"{normalized_domain} wants you to sign in with your Ethereum account:",
            f"{normalized_address}",
            "",
            statement,
            "",
            f"URI: {normalized_uri}",
            "Version: 1",
            f"Chain ID: {normalized_chain_id}",
            f"Nonce: {normalized_nonce}",
            f"Issued At: {normalized_issued_at}",
        ]
    )


def sign_siwe_message(account: LocalAccount, message: str) -> HexString:
    """
    Sign a SIWE message using EIP-191 personal_sign semantics.

    AccountingSiweAuth uses the EIP-191 prefix and ecrecover over the message bytes,
    so this must match `signMessage` / personal_sign.
    """
    signed = account.sign_message(encode_defunct(text=message))
    return f"0x{signed.signature.hex()}"


@dataclass(frozen=True)
class _CachedToken:
    token: HexString
    expires_at_ms: int


_TOKEN_TTL_MS = 23 * 60 * 60 * 1000
_TOKEN_REFRESH_SKEW_MS = 5 * 60 * 1000

_cache: dict[str, _CachedToken] = {}


@dataclass(frozen=True)
class _FailureState:
    until_ms: int
    attempts: int
    last_failure_ms: int


_FAILURE_COOLDOWN_BASE_MS = 60 * 1000
_FAILURE_COOLDOWN_MAX_MS = 5 * 60 * 1000
_FAILURE_COOLDOWN_RESET_WINDOW_MS = 10 * 60 * 1000

_failures: dict[str, _FailureState] = {}


class SiweAuthCooldownError(Exception):
    def __init__(self, *, until_ms: int, retry_after_ms: int) -> None:
        retry_after_s = max(1, (int(retry_after_ms) + 999) // 1000)
        super().__init__(
            "SIWE authentication is temporarily disabled after a recent failure. "
            f"Retry in ~{retry_after_s}s."
        )
        self.until_ms = int(until_ms)
        self.retry_after_ms = int(retry_after_ms)


def _cooldown_ms(attempts: int) -> int:
    clamped = max(1, min(int(attempts), 16))
    ms = _FAILURE_COOLDOWN_BASE_MS * (2 ** (clamped - 1))
    return int(min(ms, _FAILURE_COOLDOWN_MAX_MS))


def _record_failure(key: str, now_ms: int) -> _FailureState:
    previous = _failures.get(key)
    if previous and now_ms - previous.last_failure_ms <= _FAILURE_COOLDOWN_RESET_WINDOW_MS:
        attempts = previous.attempts + 1
    else:
        attempts = 1

    state = _FailureState(
        until_ms=now_ms + _cooldown_ms(attempts),
        attempts=attempts,
        last_failure_ms=now_ms,
    )
    _failures[key] = state
    return state


def _clear_failure(key: str) -> None:
    _failures.pop(key, None)


@dataclass
class _LoopState:
    lock: asyncio.Lock
    inflight: dict[str, asyncio.Task[HexString]]


_loop_states: weakref.WeakKeyDictionary[asyncio.AbstractEventLoop, _LoopState] = (
    weakref.WeakKeyDictionary()
)


def _get_loop_state() -> _LoopState:
    loop = asyncio.get_running_loop()
    state = _loop_states.get(loop)
    if state is None:
        state = _LoopState(lock=asyncio.Lock(), inflight={})
        _loop_states[loop] = state
    return state


def _is_expired(expires_at_ms: int, now_ms: int) -> bool:
    return expires_at_ms - now_ms <= _TOKEN_REFRESH_SKEW_MS


def _cache_key(*, cache_scope: str, chain_id: int, address: Address) -> str:
    return f"{cache_scope.rstrip('/')}:{chain_id}:{address.lower()}"


async def ensure_siwe_token(
    *,
    client: SiweAuthClient,
    chain_id: int,
    address: Address | str,
    signer: SiweSigner | LocalAccount,
    cache_scope: str | None = None,
    statement: str | None = None,
    token_ttl_ms: int = _TOKEN_TTL_MS,
) -> HexString:
    """
    Ensure the FlexvaultsClient has a valid SIWE token set for private reads.

    - Caches token in-memory keyed by {cache_scope, chain_id, address}
    - De-dupes concurrent ensure calls to avoid multiple sign flows
    """
    user = normalize_address(str(address))
    scope = (cache_scope or getattr(client, "base_url", None) or "").strip() or "flexvaults"
    key = _cache_key(cache_scope=scope, chain_id=int(chain_id), address=user)

    state = _get_loop_state()

    task: asyncio.Task[HexString]
    async with state.lock:
        now_ms = monotonic_ns() // 1_000_000
        cached = _cache.get(key)
        if cached and not _is_expired(cached.expires_at_ms, now_ms):
            client.set_siwe_token(cached.token)
            _clear_failure(key)
            return cached.token

        existing = state.inflight.get(key)
        if existing is not None:
            task = existing
        else:
            failure = _failures.get(key)
            if failure and now_ms < failure.until_ms:
                raise SiweAuthCooldownError(
                    until_ms=failure.until_ms,
                    retry_after_ms=failure.until_ms - now_ms,
                )

            async def _do() -> HexString:
                try:
                    domain_resp: SiweDomainResponse = await client.get_siwe_domain()
                    message = build_siwe_message(
                        domain=domain_resp.domain,
                        address=user,
                        chain_id=int(chain_id),
                        statement=statement or "Sign in to Flexvaults",
                    )

                    if isinstance(signer, LocalAccount):
                        signature = sign_siwe_message(signer, message)
                    else:
                        signature = normalize_hex(await signer(message))

                    login_resp: SiweLoginResponse = await client.siwe_login(
                        siwe_message=message,
                        signature=signature,
                    )
                    token = normalize_hex(login_resp.token)
                    client.set_siwe_token(token)

                    now_ms2 = monotonic_ns() // 1_000_000
                    async with state.lock:
                        _cache[key] = _CachedToken(
                            token=token, expires_at_ms=now_ms2 + int(token_ttl_ms)
                        )
                        _clear_failure(key)
                    return token
                except Exception:
                    now_ms2 = monotonic_ns() // 1_000_000
                    async with state.lock:
                        _record_failure(key, now_ms2)
                    raise

            task = asyncio.create_task(_do())
            state.inflight[key] = task

    try:
        return await task
    finally:
        async with state.lock:
            if state.inflight.get(key) is task:
                state.inflight.pop(key, None)

File: packages/sdk/src/sdk/auth/siwe.ts
'use client'

import type { WalletClient } from 'viem'
import type { Address } from '../types'

export interface SiweAuthClient {
  getSiweToken: () => string | null
  setSiweToken: (token: string) => void
  clearSiweToken: () => void
  getSiweDomain: () => Promise<{ domain: string }>
  siweLogin: (request: { siwe_message: string; signature: string }) => Promise<{ token: string }>
}

// SiweAuth tokens default to 24h validity on-chain; keep a 23h client TTL to refresh
// proactively and avoid edge-case expiry during polling/read bursts.
const TOKEN_TTL_MS = 23 * 60 * 60 * 1000
const TOKEN_REFRESH_SKEW_MS = 5 * 60 * 1000

// In-flight de-duplication to avoid multiple concurrent SIWE sign popups when several
// private-read hooks mount at once for the same account+chain.
const inflight = new Map<string, Promise<string>>()

type FailureState = {
  untilMs: number
  attempts: number
  lastFailureMs: number
}

const FAILURE_COOLDOWN_BASE_MS = 60 * 1000
const FAILURE_COOLDOWN_MAX_MS = 5 * 60 * 1000
const FAILURE_COOLDOWN_RESET_WINDOW_MS = 10 * 60 * 1000

const failures = new Map<string, FailureState>()

export class SiweAuthCooldownError extends Error {
  readonly code = 'SIWE_AUTH_COOLDOWN' as const
  readonly untilMs: number
  readonly retryAfterMs: number

  constructor(args: { untilMs: number; retryAfterMs: number }) {
    const retryAfterSeconds = Math.max(1, Math.ceil(args.retryAfterMs / 1000))
    super(
      `SIWE authentication is temporarily disabled after a recent failure. Retry in ~${retryAfterSeconds}s.`
    )
    this.name = 'SiweAuthCooldownError'
    Object.setPrototypeOf(this, SiweAuthCooldownError.prototype)
    this.untilMs = args.untilMs
    this.retryAfterMs = args.retryAfterMs
  }
}

export function isSiweAuthCooldownError(error: unknown): error is SiweAuthCooldownError {
  if (!error || typeof error !== 'object') return false
  const e = error as { code?: unknown; retryAfterMs?: unknown }
  return e.code === 'SIWE_AUTH_COOLDOWN' && typeof e.retryAfterMs === 'number'
}

function cooldownMs(attempts: number): number {
  const clampedAttempts = Math.max(1, Math.min(attempts, 16))
  const ms = FAILURE_COOLDOWN_BASE_MS * 2 ** (clampedAttempts - 1)
  return Math.min(ms, FAILURE_COOLDOWN_MAX_MS)
}

function recordFailure(storageKey: string, nowMs: number): FailureState {
  const previous = failures.get(storageKey)
  const attempts =
    previous && nowMs - previous.lastFailureMs <= FAILURE_COOLDOWN_RESET_WINDOW_MS
      ? previous.attempts + 1
      : 1
  const state: FailureState = {
    untilMs: nowMs + cooldownMs(attempts),
    attempts,
    lastFailureMs: nowMs,
  }
  failures.set(storageKey, state)
  return state
}

function clearFailure(storageKey: string): void {
  failures.delete(storageKey)
}

function getSessionStorage(): Storage | null {
  if (typeof window === 'undefined') return null
  try {
    return window.sessionStorage
  } catch {
    return null
  }
}

function safeGetItem(key: string): string | null {
  const storage = getSessionStorage()
  if (!storage) return null
  try {
    return storage.getItem(key)
  } catch {
    return null
  }
}

function safeSetItem(key: string, value: string): void {
  const storage = getSessionStorage()
  if (!storage) return
  try {
    storage.setItem(key, value)
  } catch {
    // ignore
  }
}

function safeRemoveItem(key: string): void {
  const storage = getSessionStorage()
  if (!storage) return
  try {
    storage.removeItem(key)
  } catch {
    // ignore
  }
}

function normalizeCacheScope(cacheScope?: string): string | null {
  if (!cacheScope) return null
  try {
    const url = new URL(cacheScope)
    return url.origin
  } catch {
    return cacheScope
  }
}

function getStorageKey(args: { chainId: number; address: Address; cacheScope?: string }): string {
  const scope = normalizeCacheScope(args.cacheScope)
  const addr = args.address.toLowerCase()
  return scope
    ? `flexvaults:siweToken:${scope}:${args.chainId}:${addr}`
    : `flexvaults:siweToken:${args.chainId}:${addr}`
}

type CachedToken = {
  token: string
  expiresAt: number
}

function isExpired(expiresAt: number, nowMs: number): boolean {
  return expiresAt - nowMs <= TOKEN_REFRESH_SKEW_MS
}

function parseCachedToken(raw: string): CachedToken | null {
  try {
    const parsed = JSON.parse(raw) as Partial<CachedToken>
    if (typeof parsed?.token === 'string' && typeof parsed?.expiresAt === 'number') {
      return parsed as CachedToken
    }
  } catch {
    // fall through
  }

  return null
}

function generateNonce(length = 8): string {
  const charset = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'
  const values = new Uint8Array(length)
  if (typeof globalThis.crypto?.getRandomValues !== 'function') {
    throw new Error('Secure randomness is required for SIWE nonce generation')
  }
  globalThis.crypto.getRandomValues(values)
  return Array.from(values, (v) => charset[v % charset.length]).join('')
}

function buildSiweMessage(args: {
  domain: string
  address: Address
  uri: string
  chainId: number
  statement?: string
}): string {
  const statement = args.statement ?? 'Sign in to Flexvaults'
  const nonce = generateNonce()
  const issuedAt = new Date().toISOString()

  return [
    `${args.domain} wants you to sign in with your Ethereum account:`,
    `${args.address}`,
    '',
    statement,
    '',
    `URI: ${args.uri}`,
    'Version: 1',
    `Chain ID: ${args.chainId}`,
    `Nonce: ${nonce}`,
    `Issued At: ${issuedAt}`,
  ].join('\n')
}

/**
 * Returns true if the error is a wallet user-rejection (EIP-1193 code 4001).
 * Used by polling hooks to stop retrying/polling when the user declines signing.
 */
export function isUserRejection(error: unknown): boolean {
  let current: unknown = error
  for (let depth = 0; depth < 6; depth++) {
    if (!current || typeof current !== 'object') return false
    const maybe = current as { name?: unknown; code?: unknown; cause?: unknown }

    if (maybe.name === 'UserRejectedRequestError') return true
    if (maybe.code === 4001 || maybe.code === '4001') return true

    current = maybe.cause
  }

  return false
}

/**
 * React Query refetchInterval callback factory for SIWE-authenticated polling hooks.
 * Stops polling entirely on user rejection (the user must manually refetch).
 * Pauses polling during auth cooldown, then resumes normally.
 */
export const refetchUnlessRejected =
  (pollingInterval: number) =>
  (query: { state: { error: Error | null } }): number | false => {
    const error = query.state.error
    if (isUserRejection(error)) return false
    if (isSiweAuthCooldownError(error)) {
      return Math.max(pollingInterval, error.retryAfterMs)
    }
    return pollingInterval
  }

export async function ensureSiweToken(args: {
  client: SiweAuthClient
  chainId: number
  walletClient: WalletClient
  address: Address
  /**
   * Optional namespace for token caching (e.g., API base URL). This avoids token collisions
   * if a consumer uses multiple accounting deployments on the same chain.
   */
  cacheScope?: string
}): Promise<string> {
  const storageKey = getStorageKey({
    chainId: args.chainId,
    address: args.address,
    cacheScope: args.cacheScope,
  })
  const nowMs = Date.now()

  const existing = args.client.getSiweToken()
  if (existing) {
    const cachedRaw = safeGetItem(storageKey)
    if (cachedRaw) {
      const cached = parseCachedToken(cachedRaw)
      if (cached && cached.token === existing && !isExpired(cached.expiresAt, nowMs)) {
        clearFailure(storageKey)
        return existing
      }
    }

    // In-memory token is stale unless it matches the active account/network cache key.
    // Keep sessionStorage intact so switching back to another account/network can reuse
    // its scoped cached token without forcing an unnecessary re-sign.
    args.client.clearSiweToken()
  }

  const cached = safeGetItem(storageKey)
  if (cached) {
    const parsed = parseCachedToken(cached)
    if (parsed && !isExpired(parsed.expiresAt, nowMs)) {
      args.client.setSiweToken(parsed.token)
      clearFailure(storageKey)
      return parsed.token
    }

    safeRemoveItem(storageKey)
  }

  const pending = inflight.get(storageKey)
  if (pending) return pending

  const failure = failures.get(storageKey)
  if (failure && nowMs < failure.untilMs) {
    throw new SiweAuthCooldownError({
      untilMs: failure.untilMs,
      retryAfterMs: failure.untilMs - nowMs,
    })
  }

  const promise = (async () => {
    try {
      const { domain } = await args.client.getSiweDomain()
      const chainId = args.chainId

      const message = buildSiweMessage({
        domain,
        uri: `https://${domain}`,
        chainId,
        address: args.address,
        statement: 'Sign in to Flexvaults',
      })

      const signature = await args.walletClient.signMessage({
        account: args.address,
        message,
      })

      const { token } = await args.client.siweLogin({
        siwe_message: message,
        signature,
      })

      args.client.setSiweToken(token)
      const entry: CachedToken = { token, expiresAt: Date.now() + TOKEN_TTL_MS }
      safeSetItem(storageKey, JSON.stringify(entry))
      clearFailure(storageKey)
      return token
    } catch (error) {
      if (!isUserRejection(error)) {
        recordFailure(storageKey, Date.now())
      }
      throw error
    }
  })().finally(() => inflight.delete(storageKey))

  inflight.set(storageKey, promise)
  return promise
}
